File: 2020/cpp/day3/day3_2.cpp
#include <iostream>
#include <fstream>

using namespace std;

int main()
{
    string line;

    int step_x_arr[] = {1, 3, 5, 7, 1};
    int step_y_arr[] = {1, 1, 1, 1, 2};

    unsigned int prod = 1;

    for (int i = 0; i < 5; i++)
    {
        int step_x = step_x_arr[i];
        int step_y = step_y_arr[i];
        int trees = 0;
        int index = 0;

        ifstream infile("input.txt");
        if (infile.is_open())
        {
            int n = 0;
            while (getline(infile, line))
            {
                if (n % step_y == 0)
                {
                    if (line[index] == '#')
                    {
                        trees++;
                    }
                    index += step_x;
                    if (index >= line.size())
                    {
                        index -= line.size();
                    }
                }
                n++;
            }
            infile.close();
            prod *= trees;
        }
    }
    cout << prod << endl;
}

File: 2020/python/day3/day3_2.py
#!/usr/bin/env python3
# Author: Mathias Rahbek-Borre
# Advent of code - day 3 part 2:

steps = [(1,1),(3,1),(5,1),(7,1),(1,2)]

result = None

for (i,j) in steps:
    step_x = i
    step_y = j
    index = 0
    tree_count = 0

    infile = open("input.txt", "r")
    for y, inline in enumerate(infile):
        if y % step_y == 0:
            line = inline.strip()
            if index >= len(line):
                index = index - len(line)
            if line[index] == "#":
                tree_count += 1
            index += step_x
    infile.close()
    if result == None:
        result = tree_count
    else:
        result = result*tree_count
print(result)
